src/Proxy/UploadSessionProxy.php

```diff
@@ -0,0 +1,171 @@
+				
+			case 'nextExpectedRanges':
+                return $uploadSession->getNextExpectedRanges();
+				
+			case 'uploadUrl':
+                return $uploadSession->getUploadUrl();
+			
+            default:
+                return parent::__get($name);
+        }
+    }
+	
+	/**
+     * Checks options array for key.
+     *
+     * @param string $key
+     *        The key.
+     *
+     * @return bool
+     *         The boolean value.
+     */
+	private function isOptionExists(string $key)
+	{
+		return array_key_exists($key, $this->options);
+	}
+	
+	/**
+     * Gets value corresponding to key
+	 * from options array
+     *
+     * @param string $key
+     *        The key.
+     *
+     * @return mixed
+     *         The value.
+     */
+	private function getOption(string $key)
+	{
+		return $this->options[$key];
+	}
+	
+	/**
+     * Uploads file contents as chunks.
+	 *
+     * @todo Support retry on error while uploading chunks.
+     */
+	public function run()
+	{
+		
+		$resourceStream =  $this->resource instanceof Stream ? $this->resource : Psr7\stream_for($this->resource);
+		$bytes = "";
+		$bytesRead=0;
+		$totalSize = 0;
+		
+		//Get the resource size from options array. If not available, calculate from stream contents
+		$totalSize = $this->isOptionExists("fileSize") ? $this->getOption("fileSize") : strlen($resourceStream ->getContents());
+		$resourceStream->rewind(); //rewind stream handle to start of stream
+		
+		//Set chunk size from options array. If unavailable (or) size > 60 MiB fallback to MIN_CHUNK_SIZE
+		$chunkSize = UploadSessionProxy::MIN_CHUNK_SIZE;
+		if($this->isOptionExists("chunkSize")) {
+			$chunkSize = $this->getOption("chunkSize");
+			
+			if((int)$chunkSize > UploadSessionProxy::MAX_CHUNK_SIZE){
+				$chunkSize = UploadSessionProxy::MIN_CHUNK_SIZE;
+			}
+		}
+		
+		while (!$resourceStream->eof()) {
+			
+			$bytes = $resourceStream->read($chunkSize);
+			$bytesLength = strlen($bytes);
+			
+			//initialise bytes read, and interval range
+			$byteIntervalStart=$bytesRead;
+			$bytesRead += $bytesLength;
+			$byteIntervalEnd=$bytesRead-1;
+			
+			//upload headers containing bytes Content length and range
+			$sessionHeaders = [
+				"Content-Length" => $bytesLength,
+				"Content-Range" => "bytes $byteIntervalStart-$byteIntervalEnd/$totalSize",
+			];
+			
+			//create a PUT request for chunks
+			$response = $this
+				->graph
+				->createRequest('PUT', $this->uploadUrl)
+				->addHeaders($sessionHeaders)
+				->attachBody($bytes)
+				->execute();
+			
+			$status = $response->getStatus();
+			
+			if($status != 201 && $status != 202){
+				throw new \Exception("Unexpected status code produced by 'PUT $endpoint': $status");
+			}
+			
+			if($status === 201){
+				//resource created in onedrive return response object as DriveItem
+				$resourceStream->close();
+				return $response->getResponseAsObject(DriveItem::class);
+			}
+			
+		}
+		
+		$resourceStream->close();
```

Comment:
It's risky (and inconsistent witht `upload()`) to close the stream here since we did not open it ourselves. Even when created by Guzzle (they handle it themselves in the stream cleanup code). I would not do it, and instead add to the function documentation that this responsability is left to the caller. Your help is welcome to also add a similar notice to the existing `upload()` function.